### ISSUE TITLE
Perform optional and opt-in DMN engines TCK Runners during Continuous Integration

### ISSUE THREAD

**Issue by tarilabs** [dmn-tck/tck#157]
As discussed on the 2018-06-01 meeting, I am opening this issue to track the work to allow on an optional and opt-in basis, the ability for the Continuous Integration tool (Travis CI) to perform the execution of the DMN engines TCK Runners, especially for newly raised PR.

Given the opt-in nature, for demonstrating this work I will use Camunda and Drools as example; other Vendors will be able to include their configuration or ask for it to be included, later.

I experimented with a couple options, both with their pros and cons which I describe in the following, in order for the whole group to make an informed decision when proceeding.

# Demo description
In order to experiment without touching this repository, I made [a clone](https://github.com/tarilabs/travistest-dmntck) of this repository and I created [a dummy PR](https://github.com/tarilabs/travistest-dmntck/pull/1) in order to show an example. The dummy PR demonstrate the proposal to add a new test: the DMN model is valid, however the test case xml file is intentionally broken. This is because the DMN model is composed of:
![image](https://user-images.githubusercontent.com/1699252/40879071-f19d059e-669a-11e8-8c79-35a27a5a89f1.png)
Where the decision Greetings is defined as:
![image](https://user-images.githubusercontent.com/1699252/40879078-174c7f36-669b-11e8-8bd7-d11a056e7a75.png)
The companion test xml is:
```
(...)
	<testCase id="001">
		<inputNode name="Name">
			<value>John</value>
		</inputNode>
		<resultNode name="Greetings" type="decision">
			<expected>
				<value>Hello John</value>
			</expected>
		</resultNode>
	</testCase>
	<testCase id="002">
		<inputNode name="Name">
			<value>John</value>
		</inputNode>
		<resultNode name="Greetings" type="decision">
			<expected>
				<value>Hello, John!</value> <!-- a WRONG test definition.. -->
			</expected>
		</resultNode>
	</testCase>
```
Naturally the test id `001` is valid while `002` is intentionally failing.
I am using a `002` intentionally failing test, _to simulate_ the case where a specific Vendor runner, would not cover for any reason the correctness of execution.

Hence for the purpose of this demo, you should consider this dummy PR represent a case where a new test is proposed, the test is xml valid both in the DMN model and the xml file, the test is considered correct (here simulated in the PR), and some Vendor execution does not cover the case correctly.

# Option A, use of Travis CI matrix configuration
This option uses the Travis CI matrix configuration to run the Validation as a mandatory configuration, while all the other Vendors are configured as optional/"allow-failure". In turn this means a failure in a Vendor runner, would not classify the full build as failed. To be explicit this means if the Validation (of xml) is failing, the build will be marked as Failed regardless of the Vendor runners results.

On the provided example, it would look like this:

![image](https://user-images.githubusercontent.com/1699252/40879251-93cb90fe-669d-11e8-9fc4-74c52f2c38d9.png)

And once clicking on the Details for the GitHub check would lead to the Travis CI page for this execution:

![image](https://user-images.githubusercontent.com/1699252/40879485-cc34aae0-66a0-11e8-8a49-d11d3d4d846f.png)

Being a simulation please consider the Vendor line failure, as simulating the case where the correctness of execution was not covered; because the Validation is passed, the build is marked as Success, regardless of the Vendor results.

PROs

- this is using standard Travis CI features of configuration
- the build is Success/Failure only with regards to Validation, which is the most important case anyway. This meaning we will have on the GitHub PR archive green/red only concerning the Validation phase, and regardless of any Vendor results.

CONs:

- each Vendor need to click into the Details page in order to see if their engine results for the new test as proposed in the PRs.
- The vendor execution are performed anyway, regardless of the result of the Validation.

# Option B, use of Travis CI "Build Stages" (beta-feature)
This option uses the new Beta feature of Travis CI to group jobs separately, Validation being the first group, followed by all the other Vendors runner as the second group. Currently this beta feature requires for all Groups to pass, therefore a failure in a Vendor runner, would classify the full build as failed. 

On the provided example, it would look like this:

![image](https://user-images.githubusercontent.com/1699252/40879371-2d1dc33e-669f-11e8-9e94-f1ed9ba640b5.png)

And once clicking on the Details for the GitHub check would lead to the Travis CI page for this execution:

![image](https://user-images.githubusercontent.com/1699252/40879490-e80010d4-66a0-11e8-81e4-7c8f9d20d36a.png)

Being a simulation please consider the Vendor line failure, as simulating the case where the correctness of execution was not covered; because any of the Group failed, the build is marked as Failure.

PROs

- any Vendor can immediately see on the PR page something has gone wrong, and can click into the Details page in order to see if their engine results for the new test as proposed in the PRs.
- The vendor executions should be performed only if the Validation group is passed [I haven't tested this properly]

CONs:

- this is using Beta-feature of Travis CI features of configuration
- the build is Success/Failure also with regards to any vendor results. This meaning we will have on the GitHub PR archive green/red affected by any Vendor results: we might decide we merge a PR because is a valid case, but on the GitHub archive will show as red becayse any Vendor might have failed.

# Final considerations
I preferred to show all possible options which I experimented with, by studying the Travis CI documentation for the online version we are using, despite I have naturally a preference.

My personal preference and my vote goes with Option A, because for the purpose of adding a new test, we should first most concerned with it being xml valid (for the DMN file and test xml file) and a human check the test make sense. A Vendor result should not influence with a build marked Success/Failure. Being Option A using standard features, it _should_ make more easy later to expand Travis CI usage also to produce the webpage itself.

p.s.: @falko you might want to check the Camunda runner, for the test case `002` I expect it should have reported a failure, with the input "John"
```
		<resultNode name="Greetings" type="decision">
			<expected>
				<value>Hello, John!</value> <!-- a WRONG test definition.. -->
			</expected>
		</resultNode>
```
this should be a test failure, while it is marking success instead ;)

**Comment by agilepro** [comment]
Discussed.  Need to get input from Camunda on why the runner is not reporting failures

**Comment by falko** [comment]
The default in our runner is that failing tests are reported as skipped, because we know that one test case is currently failing because of an unsupported feature and three or four others are limitations of our current runner, but not of the engine. This style of reporting is actually helpful in order to not fail the validation of a new test because of known issues with our runner or engine. Feel free to exclude our runner from the continuous integration until we changed the runner. The best solution for including the our runner in the continuous integration seems to me that we only report the known issues as skipped and fail the validation for unknown issues which then would be related to the new or changed test of the pull request.

**Comment by tarilabs** [comment]
@falko thanks for the feedback, I would like to be sure I fully understand your proposal, in order to attempt to include from the beginning all runners in the CI with confidence; how do you define:

> known issues

, please? Is there a way these _known issues_ are currently marked in the code base, can you kindly explain in a bit more detail or provide some reference/example ? Thanks!

**Comment by falko** [comment]
Known issues in the runner mean that we know from previous runner generations that the engine passes the tests, but a refactoring in the current runner makes the tests fail. The fix for these tests is unfortunately not easy with the new runner architecture.

However, we do not report success for these known issues because the current runner cannot prove it. Instead, we report skipped and unfortunately is done for the new test case in the pull request.

You can see that in the JUnit output:
[master branch test result](https://travis-ci.org/tarilabs/travistest-dmntck/jobs/387049331): Tests run: 1158, Failures: 0, Errors: 0, Skipped: **7**
[pull request test result](https://travis-ci.org/tarilabs/travistest-dmntck/jobs/387073872): Tests run: 1162, Failures: 0, Errors: 0, Skipped: **8**

Skipped tests do not affect the overall build result, that's why Travis shows the build as successful.

Let me rephrase the last sentence of my last sentence: The best solution for including our runner in the continuous integration seems to me that we only report the test cases known to fail with Camunda as skipped and fail the validation for new test cases which then would be related to the new or changed test of the pull request.

**Comment by falko** [comment]
@tarilabs I changed our runner as discussed.

**Comment by tarilabs** [comment]
Resolved with PR #166 merged.
